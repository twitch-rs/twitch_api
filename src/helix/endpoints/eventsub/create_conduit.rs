//! Creates a new conduit for your Client.
//! [`create-conduit`](https://dev.twitch.tv/docs/api/reference/#create-conduits)

use super::*;
use crate::eventsub;
use helix::RequestPost;

/// Query Parameters for [Create Conduit](super::create_conduit)
///
/// [`create-conduit`](https://dev.twitch.tv/docs/api/reference/#create-conduits)
#[derive(PartialEq, Eq, Serialize, Clone, Debug, Default)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[must_use]
#[non_exhaustive]
pub struct CreateConduitRequest {}

impl Request for CreateConduitRequest {
    type Response = eventsub::Conduit;

    const PATH: &'static str = "eventsub/conduits";
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator = twitch_oauth2::validator![];
}

/// Body Parameters for [Create Conduit](super::create_conduit)
///
/// [`create-conduit`](https://dev.twitch.tv/docs/api/reference/#create-conduits)
#[derive(PartialEq, Eq, Deserialize, Serialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[non_exhaustive]
pub struct CreateConduitBody {
    /// The number of shards to create for this conduit.
    pub shard_count: usize,
}

impl CreateConduitBody {
    /// Conduit body settings
    pub const fn new(shard_count: usize) -> Self { Self { shard_count } }
}

impl helix::private::SealedSerialize for CreateConduitBody {}

impl RequestPost for CreateConduitRequest {
    type Body = CreateConduitBody;

    fn parse_inner_response(
        request: Option<Self>,
        uri: &http::Uri,
        response: &str,
        status: http::StatusCode,
    ) -> Result<helix::Response<Self, Self::Response>, helix::HelixRequestPostError>
    where
        Self: Sized,
    {
        #[derive(PartialEq, Deserialize, Debug)]
        #[cfg_attr(feature = "deny_unknown_fields", serde(deny_unknown_fields))]
        struct InnerResponse {
            data: [eventsub::Conduit; 1],
        }

        let inner_response: InnerResponse = helix::parse_json(response, true).map_err(|e| {
            helix::HelixRequestPostError::DeserializeError(
                response.to_string(),
                e,
                uri.clone(),
                status,
            )
        })?;

        let [conduit] = inner_response.data;

        Ok(helix::Response::new(conduit, None, request, None, None))
    }
}

#[cfg(test)]
#[test]
fn test_response() {
    helix::assert_helix_snapshot!(
        CreateConduitRequest:
        req = CreateConduitRequest::default(),
        res = br#"{
            "data": [
                {
                    "id": "bfcfc993-26b1-b876-44d9-afe75a379dac",
                    "shard_count": 5
                }
            ]
        }"#,
    );
}

#[cfg(test)]
#[test]
fn test_successful_unexpected_response() {
    use helix::*;
    let req: CreateConduitRequest = CreateConduitRequest::default();

    let data = br#"{
      "data": []
    }
    "#
    .to_vec();
    let http_response = http::Response::builder().status(200).body(data).unwrap();

    let uri = req.get_uri().unwrap();
    let response = CreateConduitRequest::parse_response(Some(req), &uri, http_response);
    assert!(response.is_err());
}
