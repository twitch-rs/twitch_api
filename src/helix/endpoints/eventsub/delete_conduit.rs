//! Deletes a specified conduit.
//! [`delete-conduit`](https://dev.twitch.tv/docs/api/reference/#delete-conduit)
//!
//! Note that it may take some time for Eventsub subscriptions on a deleted conduit to show as disabled when calling [Get Eventsub Subscriptions](super::get_eventsub_subscriptions).

use super::*;
use helix::RequestDelete;

/// Query Parameters for [Delete Conduit](super::delete_conduit)
///
/// [`delete-conduit`](https://dev.twitch.tv/docs/api/reference/#delete-conduit)
#[derive(PartialEq, Eq, Serialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[must_use]
#[non_exhaustive]
pub struct DeleteConduitRequest<'a> {
    /// Conduit ID.
    #[cfg_attr(feature = "typed-builder", builder(setter(into)))]
    #[cfg_attr(feature = "deser_borrow", serde(borrow = "'a"))]
    pub id: Cow<'a, types::ConduitIdRef>,
}

impl<'a> DeleteConduitRequest<'a> {
    /// Delete a specific conduit
    pub fn new(id: impl types::IntoCow<'a, types::ConduitIdRef> + 'a) -> Self {
        Self { id: id.into_cow() }
    }
}

/// Return Values for [Delete Conduit](super::delete_conduit)
///
/// [`delete-conduit`](https://dev.twitch.tv/docs/api/reference/#delete-conduit)
#[derive(PartialEq, Eq, Deserialize, Serialize, Debug, Clone)]
#[non_exhaustive]
pub enum DeleteConduitResponse {
    /// 204 - Conduit deleted
    Success,
}

impl Request for DeleteConduitRequest<'_> {
    type Response = DeleteConduitResponse;

    const PATH: &'static str = "eventsub/conduits";
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator = twitch_oauth2::validator![];
}

impl RequestDelete for DeleteConduitRequest<'_> {
    fn parse_inner_response(
        request: Option<Self>,
        uri: &http::Uri,
        response: &str,
        status: http::StatusCode,
    ) -> Result<helix::Response<Self, Self::Response>, helix::HelixRequestDeleteError>
    where
        Self: Sized,
    {
        match status {
            http::StatusCode::NO_CONTENT | http::StatusCode::OK => Ok(helix::Response::with_data(
                DeleteConduitResponse::Success,
                request,
            )),
            _ => Err(helix::HelixRequestDeleteError::InvalidResponse {
                reason: "unexpected status",
                response: response.to_string(),
                status,
                uri: uri.clone(),
            }),
        }
    }
}

#[cfg(test)]
#[test]
fn test_response() {
    helix::assert_helix_snapshot!(
        DeleteConduitRequest:
        req = DeleteConduitRequest::new("bfcfc993-26b1-b876-44d9-afe75a379dac"),
        @r#"
    uri
    ----
    https://api.twitch.tv/helix/eventsub/conduits?id=bfcfc993-26b1-b876-44d9-afe75a379dac

    response
    ----
    Response {
        data: Success,
        pagination: None,
        request: Some(
            DeleteConduitRequest {
                id: "bfcfc993-26b1-b876-44d9-afe75a379dac",
            },
        ),
        total: None,
        other: None,
    }
    "#,
    );
}
