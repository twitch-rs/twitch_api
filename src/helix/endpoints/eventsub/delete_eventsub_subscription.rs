//! Delete an EventSub subscription.

use super::*;
use helix::RequestDelete;

/// Query Parameters for [Delete EventSub Subscriptions](super::delete_eventsub_subscription)
///
/// [`delete-eventsub-subscriptions`](https://dev.twitch.tv/docs/api/reference#delete-eventsub-subscription)
#[derive(PartialEq, Eq, Serialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[must_use]
#[non_exhaustive]
pub struct DeleteEventSubSubscriptionRequest<'a> {
    /// The subscription ID for the subscription you want to delete.
    #[cfg_attr(feature = "typed-builder", builder(setter(into)))]
    #[cfg_attr(feature = "deser_borrow", serde(borrow = "'a"))]
    pub id: Cow<'a, types::EventSubIdRef>,
}

impl<'a> DeleteEventSubSubscriptionRequest<'a> {
    /// Delete this eventsub subscription.
    pub fn id(id: impl types::IntoCow<'a, types::EventSubIdRef> + 'a) -> Self {
        Self { id: id.into_cow() }
    }
}

impl Request for DeleteEventSubSubscriptionRequest<'_> {
    type Response = DeleteEventSubSubscription;

    const PATH: &'static str = "eventsub/subscriptions";
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator = twitch_oauth2::validator![];
}

/// Return Values for [Delete EventSub Subscriptions](super::delete_eventsub_subscription)
///
/// [`delete-eventsub-subscriptions`](https://dev.twitch.tv/docs/api/reference#delete-eventsub-subscription)
#[derive(PartialEq, Eq, Deserialize, Serialize, Debug, Clone)]
#[non_exhaustive]
pub enum DeleteEventSubSubscription {
    /// 204 - Subscription deleted
    Success,
}

impl RequestDelete for DeleteEventSubSubscriptionRequest<'_> {
    fn parse_inner_response(
        request: Option<Self>,
        uri: &http::Uri,
        response: &str,
        status: http::StatusCode,
    ) -> Result<helix::Response<Self, Self::Response>, helix::HelixRequestDeleteError>
    where
        Self: Sized,
    {
        match status {
            http::StatusCode::NO_CONTENT | http::StatusCode::OK => Ok(helix::Response::with_data(
                DeleteEventSubSubscription::Success,
                request,
            )),
            _ => Err(helix::HelixRequestDeleteError::InvalidResponse {
                reason: "unexpected status",
                response: response.to_string(),
                status,
                uri: uri.clone(),
            }),
        }
    }
}

#[cfg(test)]
#[test]
fn test_request() {
    helix::assert_helix_snapshot!(
        DeleteEventSubSubscriptionRequest:
        req = DeleteEventSubSubscriptionRequest::id("deadbeef"),
    );
}
