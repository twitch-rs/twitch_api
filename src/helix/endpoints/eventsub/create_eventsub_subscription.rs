//! Creates an EventSub subscription.
//!
//! See also [`HelixClient::create_eventsub_subscription`](crate::helix::HelixClient::create_eventsub_subscription)
use super::*;
use crate::eventsub::{EventSubscription, EventType, Status, Transport, TransportResponse};

/// Query Parameters for [Create EventSub Subscription](super::create_eventsub_subscription)
///
/// [`create-eventsub-subscription`](https://dev.twitch.tv/docs/api/reference#create-eventsub-subscription)
#[derive(PartialEq, Eq, Serialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[non_exhaustive]
#[must_use]
pub struct CreateEventSubSubscriptionRequest<E: EventSubscription> {
    #[cfg_attr(feature = "typed-builder", builder(setter(skip), default))]
    #[serde(skip)]
    phantom: std::marker::PhantomData<E>,
}

impl<E: EventSubscription> CreateEventSubSubscriptionRequest<E> {
    /// Create a new eventsub subscription
    pub fn new() -> Self { Self::default() }
}

impl<E: EventSubscription> Default for CreateEventSubSubscriptionRequest<E> {
    fn default() -> Self {
        Self {
            phantom: Default::default(),
        }
    }
}

impl<E: EventSubscription> helix::Request for CreateEventSubSubscriptionRequest<E> {
    type Response = CreateEventSubSubscription<E>;

    #[cfg(feature = "twitch_oauth2")]
    const OPT_SCOPE: &'static [twitch_oauth2::Scope] = E::OPT_SCOPE;
    const PATH: &'static str = "eventsub/subscriptions";
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator = E::SCOPE;
}

/// Body Parameters for [Create EventSub Subscription](super::create_eventsub_subscription)
///
/// [`create-eventsub-subscription`](https://dev.twitch.tv/docs/api/reference#create-eventsub-subscription)
///
/// # Notes
///
/// This body is quite different from the official body. If you want the true representation in text, see [`helix::HelixRequestBody::try_to_body`] on [`CreateEventSubSubscriptionRequest<E: EventSubscription>`](CreateEventSubSubscriptionRequest)
#[derive(PartialEq, Eq, Deserialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[non_exhaustive]
pub struct CreateEventSubSubscriptionBody<E: EventSubscription> {
    /// Subscription that will be created
    #[serde(bound(deserialize = "E: EventSubscription"))]
    pub subscription: E,
    /// The notification delivery specific information
    pub transport: Transport,
}

impl<E: EventSubscription> helix::HelixRequestBody for CreateEventSubSubscriptionBody<E> {
    fn try_to_body(&self) -> Result<hyper::body::Bytes, helix::BodyError> {
        #[derive(PartialEq, Serialize, Debug)]
        struct IEventSubRequestBody<'a> {
            r#type: EventType,
            version: &'static str,
            condition: serde_json::Value,
            transport: &'a Transport,
        }

        let b = IEventSubRequestBody {
            r#type: E::EVENT_TYPE,
            version: E::VERSION,
            condition: self.subscription.condition()?,
            transport: &self.transport,
        };
        serde_json::to_vec(&b).map_err(Into::into).map(Into::into)
    }
}

// FIXME: Builder?
impl<E: EventSubscription> CreateEventSubSubscriptionBody<E> {
    /// Create a new [`CreateEventSubSubscriptionBody`]
    pub const fn new(subscription: E, transport: Transport) -> Self {
        Self {
            subscription,
            transport,
        }
    }
}

/// Return Values for [Create EventSub Subscription](super::create_eventsub_subscription)
///
/// [`create-eventsub-subscription`](https://dev.twitch.tv/docs/api/reference#create-eventsub-subscription)
#[derive(PartialEq, Eq, Deserialize, Serialize, Debug, Clone)]
#[non_exhaustive]
pub struct CreateEventSubSubscription<E: EventSubscription> {
    /// ID of the subscription created.
    pub id: types::EventSubId,
    /// Status of the subscription.
    pub status: Status,
    /// The category of the subscription that was created.
    #[serde(rename = "type")]
    pub type_: EventType,
    /// The version of the subscription type that was created.
    pub version: String,
    /// JSON object specifying custom parameters for the subscription.
    #[serde(bound(deserialize = "E: EventSubscription"))]
    pub condition: E,
    /// RFC3339 timestamp indicating when the subscription was created.
    pub created_at: types::Timestamp,
    /// JSON object indicating the notification delivery specific information. Includes the transport method and callback URL.
    pub transport: TransportResponse,
    /// Total number of subscriptions for the client ID that made the subscription creation request.
    pub total: usize,
    /// Total cost of all the subscriptions for the client ID that made the subscription creation request.
    pub total_cost: usize,
    /// The maximum total cost allowed for all of the subscriptions for the client ID that made the subscription creation request.
    pub max_total_cost: usize,
    /// How much the subscription counts against your limit.
    pub cost: usize,
}

impl<E: EventSubscription> helix::RequestPost for CreateEventSubSubscriptionRequest<E> {
    type Body = CreateEventSubSubscriptionBody<E>;

    fn parse_inner_response(
        request: Option<Self>,
        uri: &http::Uri,
        text: &str,
        status: http::StatusCode,
    ) -> Result<helix::Response<Self, Self::Response>, helix::HelixRequestPostError>
    where
        Self: Sized,
    {
        #[derive(PartialEq, Eq, Deserialize, Debug)]
        #[cfg_attr(feature = "deny_unknown_fields", serde(deny_unknown_fields))]
        pub struct InnerResponseData<E: EventSubscription> {
            cost: usize,
            #[serde(bound(deserialize = "E: EventSubscription"))]
            condition: E,
            created_at: types::Timestamp,
            id: types::EventSubId,
            status: Status,
            transport: TransportResponse,
            #[serde(rename = "type")]
            type_: EventType,
            version: String,
        }
        #[derive(PartialEq, Deserialize, Debug)]
        #[cfg_attr(feature = "deny_unknown_fields", serde(deny_unknown_fields))]
        struct InnerResponse<E: EventSubscription> {
            #[serde(bound(deserialize = "E: EventSubscription"))]
            data: Vec<InnerResponseData<E>>,
            limit: Option<usize>,
            total: usize,
            total_cost: usize,
            max_total_cost: usize,
        }
        let response: InnerResponse<E> = helix::parse_json(text, true).map_err(|e| {
            helix::HelixRequestPostError::DeserializeError(text.to_string(), e, uri.clone(), status)
        })?;
        let data = response.data.into_iter().next().ok_or_else(|| {
            helix::HelixRequestPostError::InvalidResponse {
                reason: "missing response data",
                response: text.to_string(),
                status,
                uri: uri.clone(),
            }
        })?;
        Ok(helix::Response::with_data(
            CreateEventSubSubscription {
                // helix::Response total is generally the total number of results, not what the total for this endpoint means. Thus, we set it to None.))
                total: response.total,
                total_cost: response.total_cost,
                max_total_cost: response.max_total_cost,
                cost: data.cost,
                id: data.id,
                status: data.status,
                type_: data.type_,
                version: data.version,
                condition: data.condition,
                created_at: data.created_at,
                transport: data.transport,
            },
            request,
        ))
    }
}

#[cfg(test)]
#[test]
fn test_request() {
    use crate::eventsub::{self, user::UserUpdateV1};

    let sub = UserUpdateV1::new("1234");
    let transport =
        eventsub::Transport::webhook("https://this-is-a-callback.com", "s3cre7".to_string());

    helix::assert_helix_snapshot!(CreateEventSubSubscriptionRequest::<UserUpdateV1>:
        req = CreateEventSubSubscriptionRequest::<UserUpdateV1>::default(),
        body = CreateEventSubSubscriptionBody::new(sub, transport),
        res = br#"{
            "data": [
                {
                    "id": "26b1c993-bfcf-44d9-b876-379dacafe75a",
                    "status": "webhook_callback_verification_pending",
                    "type": "user.update",
                    "version": "1",
                    "condition": {
                        "user_id": "1234"
                    },
                    "created_at": "2020-11-10T14:32:18.730260295Z",
                    "transport": {
                        "method": "webhook",
                        "callback": "https://this-is-a-callback.com"
                    },
                    "cost": 1
                }
            ],
            "total": 1,
            "total_cost": 1,
            "max_total_cost": 10000
        }"#,
        status = 202,
        @r#"
    uri
    ----
    https://api.twitch.tv/helix/eventsub/subscriptions?

    body
    ----
    {"type":"user.update","version":"1","condition":{"user_id":"1234"},"transport":{"method":"webhook","callback":"https://this-is-a-callback.com","secret":"s3cre7"}}

    response
    ----
    Response {
        data: CreateEventSubSubscription {
            id: "26b1c993-bfcf-44d9-b876-379dacafe75a",
            status: WebhookCallbackVerificationPending,
            type_: UserUpdate,
            version: "1",
            condition: UserUpdateV1 {
                user_id: "1234",
            },
            created_at: "2020-11-10T14:32:18.730260295Z",
            transport: Webhook(
                WebhookTransportResponse {
                    callback: "https://this-is-a-callback.com",
                },
            ),
            total: 1,
            total_cost: 1,
            max_total_cost: 10000,
            cost: 1,
        },
        pagination: None,
        request: Some(
            CreateEventSubSubscriptionRequest {
                phantom: PhantomData<twitch_api::eventsub::user::update::UserUpdateV1>,
            },
        ),
        total: None,
        other: None,
    }
    "#,
    );
}
