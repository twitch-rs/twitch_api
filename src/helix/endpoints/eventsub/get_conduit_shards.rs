//! Gets a list of all shards for a conduit.
//! [`get-conduit-shards`](https://dev.twitch.tv/docs/api/reference/#get-conduit-shards)

use super::*;
use crate::eventsub;
use helix::RequestGet;

/// Query Parameters for [Get Conduit Shards](super::get_conduit_shards)
///
/// [`get-conduit-shards`](https://dev.twitch.tv/docs/api/reference/#get-conduit-shards)
#[derive(PartialEq, Eq, Serialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[must_use]
#[non_exhaustive]
pub struct GetConduitShardsRequest<'a> {
    /// Conduit ID.
    #[cfg_attr(feature = "typed-builder", builder(setter(into)))]
    #[cfg_attr(feature = "deser_borrow", serde(borrow = "'a"))]
    pub conduit_id: Cow<'a, types::ConduitIdRef>,

    /// Status to filter by.
    #[cfg_attr(feature = "typed-builder", builder(default, setter(into)))]
    pub status: Option<eventsub::ShardStatus>,

    /// The cursor used to get the next page of results. The pagination object in the response contains the cursorâ€™s value.
    #[cfg_attr(feature = "typed-builder", builder(default, setter(into)))]
    #[cfg_attr(feature = "deser_borrow", serde(borrow = "'a"))]
    pub after: Option<Cow<'a, helix::CursorRef>>,
}

impl<'a> GetConduitShardsRequest<'a> {
    /// Request the shards of a conduit
    pub fn new(conduit_id: impl types::IntoCow<'a, types::ConduitIdRef> + 'a) -> Self {
        Self {
            conduit_id: conduit_id.into_cow(),
            status: None,
            after: None,
        }
    }

    /// Filter shards by a specific status
    pub const fn status(mut self, status: eventsub::ShardStatus) -> Self {
        self.status = Some(status);
        self
    }

    /// Set the cursor to get a page of results
    pub fn after(mut self, after: impl types::IntoCow<'a, helix::CursorRef> + 'a) -> Self {
        self.after = Some(after.into_cow());
        self
    }
}

/// Return Values for [Get Conduit Shards](super::get_conduit_shards)
///
/// [`get-conduit-shards`](https://dev.twitch.tv/docs/api/reference/#get-conduit-shards)
#[derive(PartialEq, Eq, Deserialize, Serialize, Debug, Clone)]
#[cfg_attr(feature = "deny_unknown_fields", serde(deny_unknown_fields))]
#[non_exhaustive]
pub struct ConduitShards {
    /// List of information about a conduit's shards.
    pub shards: Vec<eventsub::ShardResponse>,
}

impl Request for GetConduitShardsRequest<'_> {
    type Response = Vec<eventsub::ShardResponse>;

    const PATH: &'static str = "eventsub/conduits/shards";
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator = twitch_oauth2::validator![];
}

impl RequestGet for GetConduitShardsRequest<'_> {}

impl helix::Paginated for GetConduitShardsRequest<'_> {
    fn set_pagination(&mut self, cursor: Option<helix::Cursor>) {
        self.after = cursor.map(|c| c.into_cow())
    }
}

#[cfg(test)]
#[test]
fn test_request() {
    helix::assert_helix_snapshot!(
        GetConduitShardsRequest:
        req = GetConduitShardsRequest::new("12345"),
        res = br#"{
          "data": [
            {
              "id": "0",
              "status": "enabled",
              "transport": {
                "method": "webhook",
                "callback": "https://this-is-a-callback.com"
              }
            },
            {
              "id": "1",
              "status": "webhook_callback_verification_pending",
              "transport": {
                "method": "webhook",
                "callback": "https://this-is-a-callback-2.com"
              }
            },
            {
              "id": "2",
              "status": "enabled",
              "transport": {
                "method": "websocket",
                "session_id": "9fd5164a-a958-4c60-b7f4-6a7202506ca0",
                "connected_at": "2020-11-10T14:32:18.730260295Z"
              }
            },
            {
              "id": "3",
              "status": "enabled",
              "transport": {
                "method": "websocket",
                "session_id": "238b4b08-13f1-4b8f-8d31-56665a7a9d9f",
                "connected_at": "2020-11-10T14:32:18.730260295Z"
              }
            },
            {
              "id": "4",
              "status": "websocket_disconnected",
              "transport": {
                "method": "websocket",
                "session_id": "ad1c9fc3-0d99-4eb7-8a04-8608e8ff9ec9",
                "connected_at": "2020-11-10T14:32:18.730260295Z",
                "disconnected_at": "2020-11-11T14:32:18.730260295Z"
              }
            }
          ],
          "pagination": {}
        }"#,
        @r#"
    uri
    ----
    https://api.twitch.tv/helix/eventsub/conduits/shards?conduit_id=12345

    response
    ----
    Response {
        data: [
            ShardResponse {
                id: "0",
                status: Enabled,
                transport: Webhook(
                    WebhookTransportResponse {
                        callback: "https://this-is-a-callback.com",
                    },
                ),
            },
            ShardResponse {
                id: "1",
                status: WebhookCallbackVerificationPending,
                transport: Webhook(
                    WebhookTransportResponse {
                        callback: "https://this-is-a-callback-2.com",
                    },
                ),
            },
            ShardResponse {
                id: "2",
                status: Enabled,
                transport: Websocket(
                    WebsocketTransportResponse {
                        session_id: "9fd5164a-a958-4c60-b7f4-6a7202506ca0",
                        connected_at: Some(
                            "2020-11-10T14:32:18.730260295Z",
                        ),
                        disconnected_at: None,
                    },
                ),
            },
            ShardResponse {
                id: "3",
                status: Enabled,
                transport: Websocket(
                    WebsocketTransportResponse {
                        session_id: "238b4b08-13f1-4b8f-8d31-56665a7a9d9f",
                        connected_at: Some(
                            "2020-11-10T14:32:18.730260295Z",
                        ),
                        disconnected_at: None,
                    },
                ),
            },
            ShardResponse {
                id: "4",
                status: WebsocketDisconnected,
                transport: Websocket(
                    WebsocketTransportResponse {
                        session_id: "ad1c9fc3-0d99-4eb7-8a04-8608e8ff9ec9",
                        connected_at: Some(
                            "2020-11-10T14:32:18.730260295Z",
                        ),
                        disconnected_at: Some(
                            "2020-11-11T14:32:18.730260295Z",
                        ),
                    },
                ),
            },
        ],
        pagination: None,
        request: Some(
            GetConduitShardsRequest {
                conduit_id: "12345",
                status: None,
                after: None,
            },
        ),
        total: None,
        other: Some(
            {},
        ),
    }
    "#,
    );
}
