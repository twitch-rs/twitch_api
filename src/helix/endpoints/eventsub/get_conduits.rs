//! Get the conduits for your Client.
//! [`get-conduits`](https://dev.twitch.tv/docs/api/reference/#get-conduits)

use super::*;
use crate::eventsub;
use helix::RequestGet;

/// Query Parameters for [Get Conduits](super::get_conduits)
///
/// [`get-conduits`](https://dev.twitch.tv/docs/api/reference/#get-conduits)
#[derive(PartialEq, Eq, Serialize, Clone, Debug, Default)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[must_use]
#[non_exhaustive]
pub struct GetConduitsRequest {}

impl Request for GetConduitsRequest {
    type Response = Vec<eventsub::Conduit>;

    const PATH: &'static str = "eventsub/conduits";
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator = twitch_oauth2::validator![];
}

impl RequestGet for GetConduitsRequest {}

#[cfg(test)]
#[test]
fn test_request() {
    helix::assert_helix_snapshot!(
      GetConduitsRequest:
      req = GetConduitsRequest::default(),
      res = br#"{
        "data": [
          {
            "id": "26b1c993-bfcf-44d9-b876-379dacafe75a",
            "shard_count": 15
          },
          {
            "id": "bfcfc993-26b1-b876-44d9-afe75a379dac",
            "shard_count": 5
          }
        ]
      }"#,
      @r#"
    uri
    ----
    https://api.twitch.tv/helix/eventsub/conduits?

    response
    ----
    Response {
        data: [
            Conduit {
                id: "26b1c993-bfcf-44d9-b876-379dacafe75a",
                shard_count: 15,
            },
            Conduit {
                id: "bfcfc993-26b1-b876-44d9-afe75a379dac",
                shard_count: 5,
            },
        ],
        pagination: None,
        request: Some(
            GetConduitsRequest,
        ),
        total: None,
        other: Some(
            {},
        ),
    }
    "#,
    );
}
