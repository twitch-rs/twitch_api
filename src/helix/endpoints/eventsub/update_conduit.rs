//!  Updates a conduitâ€™s shard count.
//! [`update-conduit`](https://dev.twitch.tv/docs/api/reference/#update-conduits)
//!
//! To delete shards, update the count to a lower number, and the shards above the count will be deleted.
//! For example, if the existing shard count is 100, by resetting shard count to 50, shards 50-99 are disabled.

use super::*;
use crate::eventsub;
use helix::RequestPatch;

/// Query Parameters for [Update Conduit](super::update_conduit)
///
/// [`update-conduit`](https://dev.twitch.tv/docs/api/reference/#update-conduits)
#[derive(PartialEq, Eq, Serialize, Clone, Debug, Default)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[must_use]
#[non_exhaustive]
pub struct UpdateConduitRequest<'a> {
    #[serde(skip)]
    #[cfg_attr(feature = "typed-builder", builder(default))]
    _phantom: std::marker::PhantomData<&'a ()>,
}

impl Request for UpdateConduitRequest<'_> {
    type Response = eventsub::Conduit;

    const PATH: &'static str = "eventsub/conduits";
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator = twitch_oauth2::validator![];
}

/// Body Parameters for [Update Conduit](super::update_conduit)
///
/// [`update-conduit`](https://dev.twitch.tv/docs/api/reference/#update-conduits)
#[derive(PartialEq, Eq, Deserialize, Serialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[non_exhaustive]
pub struct UpdateConduitBody<'a> {
    /// Conduit ID.
    #[cfg_attr(feature = "typed-builder", builder(setter(into)))]
    #[cfg_attr(feature = "deser_borrow", serde(borrow = "'a"))]
    pub id: Cow<'a, types::ConduitIdRef>,
    /// The new number of shards for this conduit.
    pub shard_count: usize,
}

impl helix::private::SealedSerialize for UpdateConduitBody<'_> {}

impl<'a> UpdateConduitBody<'a> {
    /// Conduit body settings
    pub fn new(id: impl types::IntoCow<'a, types::ConduitIdRef> + 'a, shard_count: usize) -> Self {
        Self {
            id: id.into_cow(),
            shard_count,
        }
    }
}

impl<'a> RequestPatch for UpdateConduitRequest<'a> {
    type Body = UpdateConduitBody<'a>;

    fn parse_inner_response(
        request: Option<Self>,
        uri: &http::Uri,
        response: &str,
        status: http::StatusCode,
    ) -> Result<helix::Response<Self, Self::Response>, helix::HelixRequestPatchError>
    where
        Self: Sized,
    {
        helix::parse_single_return(request, uri, response, status)
    }
}

#[cfg(test)]
#[test]
fn test_successful_response() {
    helix::assert_helix_snapshot!(
        UpdateConduitRequest:
        req = UpdateConduitRequest::default(),
        body = UpdateConduitBody::new("bfcfc993-26b1-b876-44d9-afe75a379dac", 5),
        res = br#"{
            "data": [
                {
                    "id": "bfcfc993-26b1-b876-44d9-afe75a379dac",
                    "shard_count": 5
                }
            ]
        }"#,
        @r#"
    uri
    ----
    https://api.twitch.tv/helix/eventsub/conduits?

    body
    ----
    {"id":"bfcfc993-26b1-b876-44d9-afe75a379dac","shard_count":5}

    response
    ----
    Response {
        data: Conduit {
            id: "bfcfc993-26b1-b876-44d9-afe75a379dac",
            shard_count: 5,
        },
        pagination: None,
        request: Some(
            UpdateConduitRequest {
                _phantom: PhantomData<&()>,
            },
        ),
        total: None,
        other: None,
    }
    "#,
    );
}

#[cfg(test)]
#[test]
fn test_successful_unexpected_response() {
    use helix::*;
    let req: UpdateConduitRequest = UpdateConduitRequest::default();

    let data = br#"{
      "data": []
    }
    "#
    .to_vec();
    let http_response = http::Response::builder().status(200).body(data).unwrap();

    let uri = req.get_uri().unwrap();
    let response = UpdateConduitRequest::parse_response(Some(req), &uri, http_response);
    assert!(response.is_err());
}
