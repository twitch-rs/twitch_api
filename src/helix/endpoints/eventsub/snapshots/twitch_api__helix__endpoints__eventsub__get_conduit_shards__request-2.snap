---
source: src/helix/endpoints/eventsub/get_conduit_shards.rs
expression: response.data
---
[
    ShardResponse {
        id: "0",
        status: Enabled,
        transport: Webhook(
            WebhookTransportResponse {
                callback: "https://this-is-a-callback.com",
            },
        ),
    },
    ShardResponse {
        id: "1",
        status: WebhookCallbackVerificationPending,
        transport: Webhook(
            WebhookTransportResponse {
                callback: "https://this-is-a-callback-2.com",
            },
        ),
    },
    ShardResponse {
        id: "2",
        status: Enabled,
        transport: Websocket(
            WebsocketTransportResponse {
                session_id: "9fd5164a-a958-4c60-b7f4-6a7202506ca0",
                connected_at: Some(
                    "2020-11-10T14:32:18.730260295Z",
                ),
                disconnected_at: None,
            },
        ),
    },
    ShardResponse {
        id: "3",
        status: Enabled,
        transport: Websocket(
            WebsocketTransportResponse {
                session_id: "238b4b08-13f1-4b8f-8d31-56665a7a9d9f",
                connected_at: Some(
                    "2020-11-10T14:32:18.730260295Z",
                ),
                disconnected_at: None,
            },
        ),
    },
    ShardResponse {
        id: "4",
        status: WebsocketDisconnected,
        transport: Websocket(
            WebsocketTransportResponse {
                session_id: "ad1c9fc3-0d99-4eb7-8a04-8608e8ff9ec9",
                connected_at: Some(
                    "2020-11-10T14:32:18.730260295Z",
                ),
                disconnected_at: Some(
                    "2020-11-11T14:32:18.730260295Z",
                ),
            },
        ),
    },
]
